File: packages/client-python/src/rocketride/core/runtime/resolver.py
"""
Version resolution for runtime binaries.

Reads the compatibility range from package metadata and queries
GitHub releases to find the latest compatible runtime version.
"""

import re
from typing import Optional

import aiohttp
from packaging.specifiers import SpecifierSet
from packaging.version import Version

from ..exceptions import RuntimeNotFoundError

_GITHUB_API = 'https://api.github.com/repos/rocketride-org/rocketride-server/releases'
_TAG_PATTERN = re.compile(r'^server-v(.+)$')


def get_compat_range() -> str:
    """Read the runtime-compatible range from package metadata.

    The value is set in pyproject.toml under [tool.rocketride].
    We read it at runtime via importlib.metadata so it works in
    both installed and editable installs.
    """
    # importlib.metadata doesn't expose [tool.*] sections directly.
    # We read pyproject.toml from the package directory instead.
    from pathlib import Path

    # Walk up from this file to find pyproject.toml
    pkg_dir = Path(__file__).resolve().parent.parent.parent.parent.parent
    pyproject = pkg_dir / 'pyproject.toml'

    if pyproject.exists():
        try:
            # Use tomllib (3.11+) or tomli as fallback
            try:
                import tomllib
            except ImportError:
                import tomli as tomllib

            with open(pyproject, 'rb') as f:
                data = tomllib.load(f)

            tool = data.get('tool', {}).get('rocketride', {})
            return tool.get('runtime-compatible', '>=3.0.0,<4.0.0')
        except (KeyError, Exception):
            pass

    # Hardcoded fallback matching pyproject.toml
    return '>=3.0.0,<4.0.0'


async def resolve_compatible_version(
    compat_range: Optional[str] = None,
) -> str:
    """Query GitHub releases and return the latest version matching the compat range.

    Raises RuntimeNotFoundError if no compatible release is found.
    """
    if compat_range is None:
        compat_range = get_compat_range()

    spec = SpecifierSet(compat_range)

    async with aiohttp.ClientSession() as session:
        async with session.get(
            _GITHUB_API,
            headers={'Accept': 'application/vnd.github+json'},
            params={'per_page': 100},
        ) as resp:
            if resp.status != 200:
                raise RuntimeNotFoundError(f'Failed to query GitHub releases (HTTP {resp.status})')
            releases = await resp.json()

    candidates = []
    for release in releases:
        tag = release.get('tag_name', '')
        match = _TAG_PATTERN.match(tag)
        if not match:
            continue
        version_str = match.group(1)
        try:
            version = Version(version_str)
        except Exception:
            continue
        if version in spec:
            candidates.append(version)

    if not candidates:
        raise RuntimeNotFoundError(f'No runtime release found matching {compat_range}')

    best = max(candidates)
    return str(best)


async def resolve_docker_tag(version_spec: str) -> str:
    """Resolve a version spec to a Docker image tag.

    - ``'latest'`` -> resolve latest stable version, return ``'X.Y.Z'``
    - ``'prerelease'`` -> resolve latest prerelease version, return ``'X.Y.Z-prerelease'``
    - ``'X.Y.Z'`` -> return as-is

    The returned string is suitable for appending to the Docker image base
    (e.g. ``ghcr.io/rocketride-org/rocketride-runtime:{tag}``).
    """
    if version_spec in ('latest', 'prerelease'):
        async with aiohttp.ClientSession() as session:
            async with session.get(
                _GITHUB_API,
                headers={'Accept': 'application/vnd.github+json'},
                params={'per_page': 100},
            ) as resp:
                if resp.status != 200:
                    raise RuntimeNotFoundError(f'Failed to query GitHub releases (HTTP {resp.status})')
                releases = await resp.json()

        candidates = []
        for release in releases:
            tag = release.get('tag_name', '')
            match = _TAG_PATTERN.match(tag)
            if not match:
                continue
            version_str = match.group(1)
            is_prerelease = '-prerelease' in version_str

            if version_spec == 'prerelease' and is_prerelease:
                try:
                    candidates.append(version_str)
                except Exception:
                    continue
            elif version_spec == 'latest' and not is_prerelease:
                try:
                    candidates.append(version_str)
                except Exception:
                    continue

        if not candidates:
            raise RuntimeNotFoundError(f'No runtime release found for spec: {version_spec}')

        # Sort by PEP 440 version, taking the latest
        from .platform import _base_version

        candidates.sort(key=lambda v: Version(_base_version(v)), reverse=True)
        return candidates[0]

    # Explicit version — return as-is
    return version_spec

File: packages/client-typescript/src/client/runtime/resolver.ts
/**
 * Version resolution for runtime binaries.
 *
 * Reads the compatibility range from package metadata and queries
 * GitHub releases to find the latest compatible runtime version.
 */

import { readFileSync } from 'fs';
import { join, dirname } from 'path';
import { satisfies, maxSatisfying, valid as semverValid, compare } from 'semver';
import { RuntimeNotFoundError } from '../exceptions/index.js';
import { normalizeVersion } from './platform.js';

const GITHUB_API = 'https://api.github.com/repos/rocketride-org/rocketride-server/releases';
const TAG_PATTERN = /^server-v(.+)$/;
const DEFAULT_COMPAT = '>=3.0.0 <4.0.0';

export function getCompatRange(): string {
	// Walk up from this file to find package.json
	// Compiled: dist/cli/client/runtime/resolver.js -> 4 levels to package root
	// Source:   src/client/runtime/resolver.ts -> 3 levels to package root
	let dir = __dirname;
	for (let i = 0; i < 6; i++) {
		try {
			const pkgPath = join(dir, 'package.json');
			const raw = readFileSync(pkgPath, 'utf-8');
			const pkg = JSON.parse(raw);
			const compat: string | undefined = pkg?.rocketride?.runtimeCompatible;
			if (compat) return compat;
			break; // Found package.json but no field — use fallback
		} catch {
			dir = dirname(dir);
		}
	}
	return DEFAULT_COMPAT;
}

export async function resolveCompatibleVersion(compatRange?: string): Promise<string> {
	const range = compatRange ?? getCompatRange();

	const resp = await fetch(GITHUB_API + '?per_page=100', {
		headers: { Accept: 'application/vnd.github+json' },
	});
	if (!resp.ok) {
		throw new RuntimeNotFoundError(`Failed to query GitHub releases (HTTP ${resp.status})`);
	}
	const releases: Array<{ tag_name: string }> = await resp.json();

	const versions: string[] = [];
	for (const release of releases) {
		const match = TAG_PATTERN.exec(release.tag_name);
		if (!match) continue;
		const v = match[1];
		if (semverValid(v) && satisfies(v, range)) {
			versions.push(v);
		}
	}

	if (versions.length === 0) {
		throw new RuntimeNotFoundError(`No runtime release found matching ${range}`);
	}

	const best = maxSatisfying(versions, range);
	if (!best) {
		throw new RuntimeNotFoundError(`No runtime release found matching ${range}`);
	}
	return best;
}

/**
 * Resolve a version spec to a Docker image tag.
 *
 * - 'latest' -> latest stable (non-prerelease) from GitHub releases
 * - 'prerelease' -> latest prerelease
 * - explicit version -> return as-is
 */
export async function resolveDockerTag(versionSpec: string): Promise<string> {
	if (versionSpec !== 'latest' && versionSpec !== 'prerelease') {
		return normalizeVersion(versionSpec);
	}

	const resp = await fetch(GITHUB_API + '?per_page=100', {
		headers: { Accept: 'application/vnd.github+json' },
	});
	if (!resp.ok) {
		throw new RuntimeNotFoundError(`Failed to query GitHub releases (HTTP ${resp.status})`);
	}
	const releases: Array<{ tag_name: string }> = await resp.json();

	const candidates: string[] = [];
	for (const release of releases) {
		const match = TAG_PATTERN.exec(release.tag_name);
		if (!match) continue;
		const v = match[1];
		const isPrerelease = v.includes('-prerelease');

		if (versionSpec === 'prerelease' && isPrerelease) {
			candidates.push(v);
		} else if (versionSpec === 'latest' && !isPrerelease) {
			candidates.push(v);
		}
	}

	if (candidates.length === 0) {
		throw new RuntimeNotFoundError(`No runtime release found for spec: ${versionSpec}`);
	}

	// Sort by semver base version (strip prerelease suffix for comparison)
	candidates.sort((a, b) => {
		const baseA = a.replace(/-prerelease$/, '');
		const baseB = b.replace(/-prerelease$/, '');
		const va = semverValid(baseA);
		const vb = semverValid(baseB);
		if (va && vb) return compare(vb, va); // descending
		return 0;
	});

	return candidates[0];
}
